c++/src/kj/compat/http.c++

```diff
@@ -5716,14 +5716,111 @@ public:
+            }
+
+            kj::Promise<uint64_t> pumpTo(kj::AsyncOutputStream& output, uint64_t amount) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->pumpTo(output, amount);
+              } else {
+                return unencryptedStream->pumpTo(output, amount);
+              }
+            }
+
+            kj::Promise<void> write(const void* buffer, size_t size) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(buffer, size);
+              } else {
+                return unencryptedStream->write(buffer, size);
+              }
+            }
+
+            kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(pieces);
+              } else {
+                return unencryptedStream->write(pieces);
+              }
+            }
+
+            kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(
+                kj::AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryPumpFrom(input, amount);
+              } else {
+                return unencryptedStream->tryPumpFrom(input, amount);
+              }
+            }
+
+            kj::Promise<void> whenWriteDisconnected() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->whenWriteDisconnected();
+              } else {
+                return unencryptedStream->whenWriteDisconnected();
+              }
+            }
+
+            void shutdownWrite() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->shutdownWrite();
+              } else {
+              unencryptedStream->shutdownWrite();
+              }
+            }
+
+            void abortRead() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->abortRead();
+              } else {
+                unencryptedStream->abortRead();
+              }
+            }
+
+            kj::Maybe<int> getFd() const override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->getFd();
+              } else {
+                return unencryptedStream->getFd();
+              }
+            }
+
+            void startTls(
+                kj::SecureNetworkWrapper* wrapper, kj::StringPtr expectedServerHostname) {
+              kj::Promise<kj::Own<kj::AsyncIoStream>> secureStream =
+                  wrapper->wrapClient(kj::mv(unencryptedStream), expectedServerHostname);
+              encryptedStream = kj::newPromisedStream(kj::mv(secureStream));
+            }
+
+          private:
+            kj::Own<kj::AsyncIoStream> unencryptedStream;
+            kj::Maybe<kj::Own<kj::AsyncIoStream>> encryptedStream;
+          };
+
+          auto refConnection = kj::refcountedWrapper(
```

Comment:
I don't think you really need reference counting here. It should be understood that the starter function cannot be called if the stream has already been discarded. So the function lambda only needs to capture a reference to the object, it doesn't need a strong refcount.